python_modules/dagster/dagster/_core/definitions/sensor_definition.py

```diff
@@ -271,28 +313,222 @@ def latest_materialization_records_by_key(
+            raise DagsterInvalidInvocationError(
+                f"Asset key {asset_key} not monitored in sensor definition"
+            )
+
+        partitions_def = self._partitions_def_by_asset_key.get(asset_key)
+
+        materialization_by_partition: Dict[str, Optional[EventLogRecord]] = {}
+        if partitions_def == None:
+            raise DagsterInvariantViolationError(
+                "Cannot get latest materialization by partition for assets with no partitions"
+            )
+        else:
+            _, cursor = self._get_cursor(asset_key)
+
+            partitions_to_fetch = (
+                self._get_partitions_after_cursor(asset_key)
+                if after_cursor_partition
+                else list(partitions_def.get_partition_keys())
+            )
+
+            partition_materializations = self.instance.get_event_records(
+                EventRecordsFilter(
+                    event_type=DagsterEventType.ASSET_MATERIALIZATION,
+                    asset_key=asset_key,
+                    asset_partitions=partitions_to_fetch,
+                    after_cursor=cursor,
+                ),
+            )
+            for partition in partitions_to_fetch:
+                materialization_by_partition[partition] = None
+            for materialization in partition_materializations:
+                # Only update if no previous materialization for this partition exists,
+                # to ensure we return the most recent materialization for each partition.
+                partition = self.get_partition_from_event_log_record(materialization)
+                if materialization_by_partition[partition] is None:
+                    materialization_by_partition[partition] = materialization
+
+        return materialization_by_partition
+
+    @public
+    def get_cursor_partition(self, asset_key):
+        """A utility method to get the current partition the cursor is on."""
+        asset_key = check.opt_inst_param(asset_key, "asset_key", AssetKey)
+        if asset_key not in self._asset_keys:
+            raise DagsterInvalidInvocationError(
+                "Provided asset key must correspond to a provided asset"
+            )
+        if asset_key:
+            partition_key, _ = self._get_cursor(asset_key)
+        elif self._asset_keys is not None and len(self._asset_keys) == 1:
+            partition_key, _ = self._get_cursor(self._asset_keys[0])
+        else:
+            raise DagsterInvalidInvocationError(
+                "Asset key must be provided when multiple assets are defined"
+            )
+
+        return partition_key
+
+    def _can_map(self, to_partitions: PartitionsDefinition, from_partitions: PartitionsDefinition):
+        if isinstance(to_partitions, TimeWindowPartitionsDefinition) and isinstance(
+            from_partitions, TimeWindowPartitionsDefinition
+        ):
+            return True
+        return False
+
+    @public
+    def map_partition(
```

Comment:
A couple thoughts on the signature of this method:
- A name that indicates what it returns would be most clear. E.g. `get_downstream_partition_keys` or something?
- It might save the user some characters if it accepts an `AssetKey` instead of a `PartitionsDefinition`?

Reply:
Yep, I can change the name of the method!

Agree that it is easier to accept an asset key. But there's a situation that this might be a little awkward. So a user might pass in the partitions definition of the job the sensor yields a run request for. For example:

```
@multi_asset_sensor(asset_keys=[AssetKey("a")], job_name="asset_b_and_c_job")
def my_sensor(context):
   ...
   context.map_partition(partition_key, to_partitions=asset_b_and_c_job.partitions_def)
```
We could also allow the sensor to have access to the partitions definition of the targetted job, but it seems a little odd to have to select one asset out of the job. I guess another option is to add an additional param `from_job_partitions`?